File: src/Core/Camera.ts
import { Vector3 } from './Math';
import { Matrix4 } from './Math';

export class Camera {
    public position: Vector3;
    public target: Vector3;
    public up: Vector3;
    public fov: number;
    public aspect: number;
    public near: number;
    public far: number;

    private viewMatrix: Matrix4;
    private projectionMatrix: Matrix4;

    constructor(
        position: Vector3 = new Vector3(0, 0, 5),
        target: Vector3 = Vector3.zero(),
        fov: number = 45 * Math.PI / 180,
        aspect: number = 1,
        near: number = 0.1,
        far: number = 100
    ) {
        this.position = position;
        this.target = target;
        this.up = new Vector3(0, 1, 0);
        this.fov = fov;
        this.aspect = aspect;
        this.near = near;
        this.far = far;

        this.viewMatrix = new Matrix4();
        this.projectionMatrix = new Matrix4();
        this.updateMatrices();
    }

    updateMatrices(): void {
        this.projectionMatrix.perspective(this.fov, this.aspect, this.near, this.far);

        this.viewMatrix.lookAt(
            new Float32Array([this.position.x, this.position.y, this.position.z]),
            new Float32Array([this.target.x, this.target.y, this.target.z]),
            new Float32Array([this.up.x, this.up.y, this.up.z])
        );
    }

    getViewMatrix(): Matrix4 {
        return this.viewMatrix;
    }

    getProjectionMatrix(): Matrix4 {
        return this.projectionMatrix;
    }

    setAspect(aspect: number): void {
        this.aspect = aspect;
        this.updateMatrices();
    }

    lookAt(target: Vector3): void {
        this.target = target;
        this.updateMatrices();
    }

    setPosition(position: Vector3): void {
        this.position = position;
        this.updateMatrices();
    }

    orbit(deltaX: number, deltaY: number, distance: number): void {
        const theta = deltaX * 0.01;
        const phi = deltaY * 0.01;

        const x = distance * Math.sin(phi) * Math.cos(theta);
        const y = distance * Math.cos(phi);
        const z = distance * Math.sin(phi) * Math.sin(theta);

        this.position = new Vector3(x, y, z);
        this.updateMatrices();
    }
}

File: VulkanEngine/Core/Entities/Camera.cs
using VulkanEngine.Core.Math;

namespace VulkanEngine.Core.Entities;

public class Camera
{
    public Vector3 Position { get; set; }
    public Vector3 Target { get; set; }
    public Vector3 Up { get; set; }
    public float Fov { get; set; }
    public float AspectRatio { get; set; }
    public float Near { get; set; }
    public float Far { get; set; }

    public Camera(
        Vector3? position = null,
        Vector3? target = null,
        float fov = 45f,
        float aspectRatio = 16f / 9f,
        float near = 0.1f,
        float far = 100f)
    {
        Position = position ?? new Vector3(0, 2, 5);
        Target = target ?? Vector3.Zero;
        Up = Vector3.Up;
        Fov = fov * MathF.PI / 180f;
        AspectRatio = aspectRatio;
        Near = near;
        Far = far;
    }

    public Vector3 Forward => (Target - Position).Normalized;
    public Vector3 Right => Vector3.Cross(Forward, Up).Normalized;

    public void LookAt(Vector3 target)
    {
        Target = target;
    }

    public void SetPosition(Vector3 position)
    {
        Position = position;
    }

    public void SetAspectRatio(float aspectRatio)
    {
        AspectRatio = aspectRatio;
    }

    public void Orbit(float deltaX, float deltaY, float distance)
    {
        var theta = deltaX * 0.01f;
        var phi = deltaY * 0.01f;

        var x = distance * MathF.Sin(phi) * MathF.Cos(theta);
        var y = distance * MathF.Cos(phi);
        var z = distance * MathF.Sin(phi) * MathF.Sin(theta);

        Position = new Vector3(x, y, z);
    }

    public void Move(Vector3 direction, float speed)
    {
        Position += direction * speed;
    }

    public void MoveForward(float speed) => Move(Forward, speed);
    public void MoveBackward(float speed) => Move(-Forward, speed);
    public void MoveLeft(float speed) => Move(-Right, speed);
    public void MoveRight(float speed) => Move(Right, speed);
    public void MoveUp(float speed) => Move(Up, speed);
    public void MoveDown(float speed) => Move(-Up, speed);
}
